### PR TITLE
[s11n 6/7] Serialisation test

### PR DESCRIPTION
**Pull request by jarlamsa**
## Description

This is a side-port to development of PR https://github.com/ARMmbed/mbedtls-restricted/pull/575 that has already been merged to baremetal.

**Dependencies:** based on #2764, #2766 #2785 #2788 and #2789  - see "Background" below.

This PR adds ssl_client2/ssl_server2 -tests for the session serialisation/deserialisation.

## Background

This is the sixth in a series of seven PRs implementing serialization of a full SSL context (not just the session), under some restrictions.

The PRs in the series are:
 
1. [r563](https://github.com/ARMmbed/mbedtls-restricted/pull/563) -> #2764 Refactor derive keys
2. [r568](https://github.com/ARMmbed/mbedtls-restricted/pull/568) -> #2766 Make session serialisation functions public 
3. [r570](https://github.com/ARMmbed/mbedtls-restricted/pull/570) -> #2785 Make session format ABI-independent
4. [r567](https://github.com/ARMmbed/mbedtls-restricted/pull/567) -> #2788 Add version + config check to SSL tickets
5. [r572](https://github.com/ARMmbed/mbedtls-restricted/pull/572) -> #2789 Context serialisation: API and documentation
6. [r575](https://github.com/ARMmbed/mbedtls-restricted/pull/575) -> #2791 Serialisation test __(this PR)__
7. [r616](https://github.com/ARMmbed/mbedtls-restricted/pull/616) -> #2792 Implement context serialisation

The full series is already merged in baremetal. It needs upstreaming for the benefit of the Could team, who will be the main user, and relies on upstream releases.

## Side-porting details

Side-porting was done with `git cherry-pick`; all commits were picked with no conflicts.

The table shows how each side-ported commit was handled:

- (nothing) cherry-picking worked without conflict
- "t" only trivial context-related conflicts (eg. independent additions to the same place)
- "r" manual rework after cherry-picking
- "R" the commit was reconstructed manually (moving code around)
- "+" added in this side-port PR

| resolution | original commit | commit message |
|---| ---          | --- |
| t | 0ea3cfe5bf2f | Add option for serialization in ssl_client/server2 |
| | d736d086184d | Serialization/deserialization in ssl_client2 |
| | 654e8de16302 | Rely on opt.exchanges for sending after serialization |
| | f457293b5791 | Serialize/deserialize for ssl_server2 |
| | cf1b6724c968 | Use MBEDTLS_SSL_CONTEXT_SERIALIZATION flag |
| | fa45e6005e9a | Add serialization tests to ssl-opt.sh |
| | dcfc2a736455 | Add missing slashes to tests |
| | cc281b8a549a | ssl-opt.sh tests for serialization are currently using stub implement… |
| | 38061f4a179d | Remove mbedtls_ssl_free() and mbedtls_ssl_init() from serialization f… |
| | 5737ec97a620 | Address review comments for code-style issues |
| | f4f8ed7e82ab | Allow stub implementation of the context_save for now |
| | bff4a91827eb | Fix spacing |
| | b5ff6a4ced96 | Add option for ssl-context re-initialization flow |
| | 8a91c0650a96 | Add tests for re-init flow for context serialization |
| | 034ae84601cc | Fix compiler warnings |
| | 85c238018e03 | Fix spacing |
| | 29a15c2de528 | Set timer callbacks with serialization |
| | 8b2608b6c2a5 | Fix style issues |

## Review notes

Reviewers are strongly encouraged to review commit by commit, as the final diff is unlikely to make sense. Please note that the first commits (up to "Clarify documentation of mbedtls_ssl_context_load()") are part of #2764, #2766 #2785 #2788 and #2789 and don't need to be re-reviewed here (except checking that they're indeed part of #2764 #2766 #2785 #2788 #2789).

### CODE REVIEW

**Comment by jarlamsa** [comment]
Rebased on top of #2789

**Comment by jarlamsa** [comment]
@hanno-arm please re-review

File: programs/ssl/ssl_client2.c [hunk 1]
```diff
@@ -2844,6 +2844,11 @@ int main( int argc, char *argv[] )
                                     mbedtls_net_send, mbedtls_net_recv,
                                     opt.nbio == 0 ? mbedtls_net_recv_timeout : NULL );
 
+#if defined(MBEDTLS_TIMING_C)
+            if( opt.nbio != 0 && opt.read_timeout != 0 )
+            mbedtls_ssl_set_timer_cb( &ssl, &timer, mbedtls_timing_set_delay,
```
Comment (artokin) [inline programs/ssl/ssl_client2.c]: Bad indentation

File: programs/ssl/ssl_server2.c [hunk 2]
```diff
@@ -3852,6 +3852,11 @@ int main( int argc, char *argv[] )
                 mbedtls_ssl_set_bio( &ssl, &client_fd, mbedtls_net_send, mbedtls_net_recv,
                                     opt.nbio == 0 ? mbedtls_net_recv_timeout : NULL );
 
+#if defined(MBEDTLS_TIMING_C)
+            if( opt.nbio != 0 && opt.read_timeout != 0 )
+            mbedtls_ssl_set_timer_cb( &ssl, &timer, mbedtls_timing_set_delay,
```
Comment (artokin) [inline programs/ssl/ssl_server2.c]: Bad indentation

**Comment by jarlamsa** [comment]
Thank you for the review @artokin, the indentation was fixed in https://github.com/ARMmbed/mbedtls/commit/8b2608b6c2a5368ff09f35e41d294fd4de592be0

**Comment by jarlamsa** [comment]
Rebased without conflicts, then amended "Add option for serialization in ssl_client/server2" and fixed an issue with changed function names.

**Comment by jarlamsa** [comment]
@hanno-arm @artokin please re-review

**Comment by jarlamsa** [comment]
All of the serialisation PRs now have 2 approving reviews and passing CI. Thus considering all of them ready for merge.

**Pull request merged**